src/test/fuzz/block_index.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chain.h>
+#include <chainparams.h>
+#include <txdb.h>
+#include <validation.h>
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+#include <test/util/setup_common.h>
+
+namespace {
+// Hardcoded block hash and nBits to make sure the blocks we store pass the pow check.
+const uint256 g_block_hash{uint256S("000000002c05cc2e78923c34df87fd108b22221ac6076c18f3ade378a4d915e9")};
+uint32_t g_bits{0x1d00ffff};
+
+util::SignalInterrupt g_interrupt;
+
+//! Compare two block files.
+bool BlockFileEq(const CBlockFileInfo& a, const CBlockFileInfo& b)
+{
+    return a.nBlocks == b.nBlocks &&
+        a.nSize == b.nSize &&
+        a.nUndoSize == b.nUndoSize &&
+        a.nHeightFirst == b.nHeightFirst &&
+        a.nHeightLast == b.nHeightLast &&
+        a.nTimeFirst == b.nTimeFirst &&
+        a.nTimeLast == b.nTimeLast;
+}
+
+} // namespace
+
+void init_block_index()
+{
+    SelectParams(ChainType::MAIN);
+}
+
+FUZZ_TARGET(block_index, .init = init_block_index)
+{
+    FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
+    auto block_index = CBlockTreeDB(DBParams{
+        .path = "", // Memory only.
+        .cache_bytes = 1 << 20, // 1MB.
+        .memory_only = true,
+    });
+
+    // Generate a number of block files to be stored in the index.
+    int files_count = fuzzed_data_provider.ConsumeIntegralInRange(1, 100);
+    std::vector<std::unique_ptr<CBlockFileInfo>> files;
+    files.reserve(files_count);
+    std::vector<std::pair<int, const CBlockFileInfo*>> files_info;
+    files_info.reserve(files_count);
+    for (int i = 0; i < files_count; i++) {
+        if (auto file_info = ConsumeDeserializable<CBlockFileInfo>(fuzzed_data_provider)) {
```

Comment:
In 8a0cb8b2147e852ac80d2030057272bdb59a83f2: Instead of using `ConsumeDeserializable`, couldn't we have a function to create a `CBlockFileInfo`? E.g.:

```diff
diff --git a/src/test/fuzz/block_index.cpp b/src/test/fuzz/block_index.cpp
index b5b25fcbc7..528c2fae9f 100644
--- a/src/test/fuzz/block_index.cpp
+++ b/src/test/fuzz/block_index.cpp
@@ -37,6 +37,19 @@ void init_block_index()
     SelectParams(ChainType::MAIN);
 }
 
+CBlockFileInfo CreateCBlockFileInfo(FuzzedDataProvider& fuzzed_data_provider)
+{
+    CBlockFileInfo block_file_info;
+    block_file_info.nBlocks = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+    block_file_info.nSize = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+    block_file_info.nUndoSize = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+    block_file_info.nHeightFirst = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+    block_file_info.nHeightLast = fuzzed_data_provider.ConsumeIntegral<unsigned int>();
+    block_file_info.nTimeFirst = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
+    block_file_info.nTimeLast = fuzzed_data_provider.ConsumeIntegral<uint64_t>();
+    return block_file_info;
+}
+
 FUZZ_TARGET(block_index, .init = init_block_index)
 {
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};
@@ -53,12 +66,9 @@ FUZZ_TARGET(block_index, .init = init_block_index)
     std::vector<std::pair<int, const CBlockFileInfo*>> files_info;
     files_info.reserve(files_count);
     for (int i = 0; i < files_count; i++) {
-        if (auto file_info = ConsumeDeserializable<CBlockFileInfo>(fuzzed_data_provider)) {
-            files.push_back(std::make_unique<CBlockFileInfo>(std::move(*file_info)));
-            files_info.push_back({i, files.back().get()});
-        } else {
-            return;
-        }
+        auto file_info{CreateCBlockFileInfo(fuzzed_data_provider)};
+        files.push_back(std::make_unique<CBlockFileInfo>(std::move(file_info)));
+        files_info.push_back({i, files.back().get()});
     }
 
     // Generate a number of block headers to be stored in the index.
```

Reply:
Good call, i've adopted this approach for both the `CBlockFile`s and the `CBlockHeader`s.

Reply:
My thinking was that it would actually be more efficient to not rely on `ConsumeDeserializable` which needs to first guess the length of the byte vector to be consumed:
https://github.com/bitcoin/bitcoin/blob/a4ca4975880c4f870c6047065c70610af2529e74/src/test/fuzz/util.h#L95-L107

Reply:
+1 

> My thinking was that it would actually be more efficient to not rely on ConsumeDeserializable which needs to first guess the length of the byte vector to be consumed

My initial thought was it facilitates especially for cases that `files_count`/`files_count` is closer to their max possible value.

Reply:
Alright, in order to benchmark which approach was most efficient i ran both versions with `-runs=100000` on an empty folder 3 times and compared the average coverage and runtime.

- `ConsumeIntegral`-based: runtime of 75 seconds for a coverage of `1565` for all 3 runs.
- `ConsumeDeserializable`-based: average runtime of 12 seconds for a coverage of `1560`.

Given the burst in coverage at the start of the run, the clear difference in runtime and the small difference in coverage i figured i'd need better measurement. I ran both on an empty folders with `-runs=1000000`:
- `ConsumeIntegral`-based: coverage of `2230` after 1M runs (runtime: 2872 seconds).
- `ConsumeDeserializable`-based: coverage of `2167` after 1M runs (runtime: 1614 seconds).

My interpretation of these results is that the `ConsumeDeserializable`-based target runs faster because of the invalid deserializations, that both targets quickly get to some basic coverage, but that the `ConsumeIntegral`-based one eventually gets to produce more interesting coverage. In my opinion i should therefore keep the `ConsumeIntegral`-based version even though its coverage/second is lower.

Reply:
Yea, `ConsumeDeserializable` is faster because it will 'return' every time it gets an invalid deserialization. The coverage is similar but I believe that it's expected because the difference between both approaches will only reflect on the size of blocks and files.

Reply:
I've reverted back to using `ConsumeDeserializable`. I had the same intuition as @brunoerg but from my testing it's not clear that the custom way is more efficient (in terms of coverage per unit of time). Absent this, it makes sense to not introduce more code and simply use the existing utilities.